File: modules/newpage.py
import json

from modules import ModuleError
from renderer import RenderContext, render_template_from_string
from urllib.parse import quote
from web.controllers import articles


def has_content():
    return False


def allow_api():
    return True


def render(context: RenderContext, params):
    example = params.get('example', '')
    category = params.get('category', '')
    submit_text = params.get('submit', '创建页面')

    placeholder = f"比如说, {example}" if example else "比如说, new-page-1"

    config = json.dumps({'category': category})

    return render_template_from_string("""
    <div class="new-page-form w-newpage-module" data-config="{{ config }}">
      <form method="get" action="">
        <p>
          <div>页面名称:</div>
          <input name="new_fullname" type="text" placeholder="{{ placeholder }}" required="true">
        </p>
        <p>
          <input value="{{ submit_text }}" type="submit">
        </p>
      </form>
    </div>
    """,
    placeholder=placeholder,
    config=config,
    submit_text=submit_text
    )


def api_check(context, params):
    new_fullname = (params.get('new_fullname') or '').strip()
    category = (params.get('category') or '').strip()

    if not new_fullname:
        raise ModuleError('请输入页面名称')

    if category:
        full_name = f"{category}:{new_fullname}"
    else:
        full_name = new_fullname

    full_name = articles.normalize_article_name(full_name)

    if not articles.is_full_name_allowed(full_name):
        raise ModuleError('此页面名称不可用')

    existing = articles.get_article(full_name)
    if existing is not None:
        raise ModuleError('此页面已存在')

    return {'url': f"/{quote(full_name, safe=':')}/edit/true"}

File: internal/modules/newpage.go
package modules

import (
	"github.com/WikitTeam/ProjectWikit/internal/escape"
	"github.com/WikitTeam/ProjectWikit/internal/module"
	"github.com/WikitTeam/ProjectWikit/internal/wikijson"
)

func init() { module.Register("newpage", renderNewPage) }

func renderNewPage(env module.Env, params map[string]string, _ string) (string, error) {
	// An empty submit= is the caller asking for a blank button, so a missing
	// key and a present one are not the same thing here.
	submit, ok := params["submit"]
	if !ok {
		submit = env.Text("module-newpage-submit")
	}
	config, err := wikijson.Marshal(wikijson.Object{{Key: "category", Value: params["category"]}})
	if err != nil {
		return "", err
	}

	// new-page-box appears nowhere else in the tree because the base stylesheet
	// reserves it for exactly this form.
	return `<div class="new-page-box new-page-form w-newpage-module" data-config="` + escape.HTML(config) + `">
  <form method="get" action="">
    <input class="text" name="new_fullname" type="text" size="30" placeholder="` +
		escape.HTML(params["example"]) + `" required="true">
    <input class="button" value="` + escape.HTML(submit) + `" type="submit">
  </form>
</div>`, nil
}
